packages/financial-templates-lib/src/clients/InsuredBridgeL2Client.ts

```diff
@@ -122,14 +130,119 @@ export class InsuredBridgeL2Client {
+      // TODO: Do something with eventsData.missingEvents array?
+      throw new Error(
+        `${eventsData.missingEvents.length} WhitelistToken events found in fallback l2 provider not found in all l2 web3 providers`
+      );
+    }
+
+    // All events were found in all providers, can return any of the event data arrays
+    return eventsData.events;
   }
 
+  /**
+   * Fetches specified contract event data for all input web3 providers. Returns false if any of the events found with
+   * one provider are NOT matched exactly in ALL of the other providers' event arrays.
+   * @param web3s Web3 providers to check for target event.
+   * @param contractAbi Contract ABI to query target event on.
+   * @param contractAddress Contract address to query target event on.
+   * @param eventName Name of target event.
+   * @param eventSearchOptions Target event search options. See here for more details: https://web3js.readthedocs.io/en/v1.5.2/web3-eth-contract.html#getpastevents
+   * @returns Object containing success of event query, missing events not found in all providers, and all event data
+   */
+  async getEventsForMultipleProviders(
+    web3s: Web3[],
+    contractAbi: any[],
+    contractAddress: string,
+    eventName: string,
+    eventSearchOptions: EventSearchOptions
+  ): Promise<{ missingEvents: EventData[]; events: EventData[] }> {
+    const allProviderEvents = await Promise.all(
+      web3s.map((_web3) => {
+        const _contract = new _web3.eth.Contract(contractAbi, contractAddress);
+        return _contract.getPastEvents(eventName, eventSearchOptions);
+      })
+    );
+
+    // Associate each event with a key uniquely identifying the event. We'll use this key in the next step to determine
+    // which events were returned by all providers.
+    type UniqueEventData = { [uniqueEventKey: string]: EventData };
+    const uniqueEventsForProvider: UniqueEventData[] = [];
+    // [index of web3Provider => {eventKey => event}]
+    const uniqueEvents: UniqueEventData = {};
+    // Union map of ALL unique events across all providers: [{eventKey => event}]. We'll use this map as the starting
+    // point for all of the matched events that we'll return from this function.
+
+    const _getUniqueEventKey = (event: EventData): string => {
+      return JSON.stringify({
+        transactionHash: event.transactionHash,
+        transactionIndex: event.transactionIndex,
+        logIndex: event.logIndex,
+        returnValues: event.returnValues,
+        address: event.address,
+      });
+    };
+    allProviderEvents.forEach((eventDataForProvider, i) => {
+      uniqueEventsForProvider[i] = {};
+      eventDataForProvider.forEach((event) => {
+        const uniqueEventKey = _getUniqueEventKey(event);
+        // Add event for this provider.
+        uniqueEventsForProvider[i][uniqueEventKey] = event;
+        // Add event to union map if we haven't seen it before.
+        uniqueEvents[uniqueEventKey] = event;
```

Comment:
i think this would be useful to also store a map of  counts, and increment it every time you run into an event with that key.  This way u can just iterate over the counts in the following loop and it might simplify some of the logic there

Reply:
But you'd need to remember which provider index saw that event?

Reply:
i didnt see where that mattered, it looks like u just care if the event exists and if its found in all providers. you already have the map of uniqueEvents to do a lookup without having to know which provider it was from.  i might be missing something though

Reply:
Ah I see what you're saying, this is a really clever suggestion! PTAL at [this implementation](https://github.com/UMAprotocol/protocol/pull/3653/commits/f6e5c58c8e8d6681e9a1320332385624186006e6)